### DISCUSSION TITLE
How to trace mysql database calls

### DISCUSSION THREAD

**Discussion by kaykhan** [aws-powertools/powertools-lambda-typescript#2027] [Q&A] [answered]
I have a lambda function making database calls to aws rds mysql. Currently im using the prisma orm. 

I saw in the aws xray documentation there is a way to capture mysql https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-nodejs-sqlclients.html

- How is this done using powertools-lambda-typescript library?
- Do i need to bring in xray for this? If i import both xray and powertools do they play nicely with one another?

Ive tried using the escape hatch  https://docs.powertools.aws.dev/lambda/typescript/latest/core/tracer/#escape-hatch-mechanism

But there is no captureMySql on the .provider

![image](https://github.com/aws-powertools/powertools-lambda-typescript/assets/5202527/5ad2579f-8888-4f51-b7d7-ca3da6eca71a)

        "@aws-lambda-powertools/tracer": "^1.17.0",

**Comment by dreamorosi** [comment, marked as answer]
Hi, at the moment the Tracer does not support that X-Ray feature, we have a feature request (#749) that has been open for a while but hasn't gotten a lot of traction (so far at least).

The reason for this not being available in Powertools is that to keep bundle size and performance in check we are using only the `aws-xray-sdk-core` package ([npm](https://www.npmjs.com/package/aws-xray-sdk-core)) while the features you've read about are available as a separate package `aws-xray-sdk-mysql` ([npm](https://www.npmjs.com/package/aws-xray-sdk-mysql)).

Given our [project's tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets), specifically the "Keep it lean" one, we usually add additional dependencies only whenever strictly necessary. This usually means whenever there's enough demand.

As a first step, I'd encourage you to chime in on the issue I linked (#749) as well as reacting with a 👍 on the original message. Every quarter or so we re-evaluate our [project's roadmap](https://docs.powertools.aws.dev/lambda/typescript/latest/roadmap/) and in most cases we rank feature requests based on how many customers are interested in them.

As an alternative solution, and until the feature is available (if ever), you could still trace the queries by creating a custom segment and annotating it with the query you're running, for example:

```ts
import { Tracer } from '@aws-lambda-powertools/tracer';

const tracer = new Tracer({ serviceName: 'serverlessAirline' });

const mySqlClient = {}

const runTracedQuery = async (query: unknown): Promise<unknown> => {
  const parentSubsegment = tracer.getSegment(); // This is the subsegment currently active
  let subsegment;
  if (parentSubsegment) {
    // Create subsegment for the function & set it as active
    subsegment = parentSubsegment.addNewSubsegment(`### chargeId`);
    tracer.setSegment(subsegment);
  }

  // add the query as metadata
  tracer.putMetadata('query', query);

  let res;
  try {
    /* run your query here using your client */
    res = await mySqlClient(query);

    // Add the response as metadata
    tracer.addResponseAsMetadata(res, 'queryResult');
  } catch (err) {
    // Add the error as metadata
    tracer.addErrorAsMetadata(err as Error);
    throw err;
  }

  if (parentSubsegment && subsegment) {
    // Close subsegment (the AWS Lambda one is closed automatically)
    subsegment.close();
    // Set the facade segment as active again
    tracer.setSegment(parentSubsegment);
  }

  return res;
};
```

Note that I have not tested the code above, but it's a refactoring of the snippets found [here](https://docs.powertools.aws.dev/lambda/typescript/latest/core/tracer/#__tabbed_5_2) and [here](https://docs.powertools.aws.dev/lambda/typescript/latest/core/tracer/#__tabbed_4_2).

Hope it helps!

**Reply by dreamorosi** [reply]
For future readers, it's also possible to use the `aws-xray-sdk-mysql` alongside Tracer to capture queries to the database, the code would look something like this:

```ts
import { Logger } from '@aws-lambda-powertools/logger';
import { Tracer } from '@aws-lambda-powertools/tracer';
import { captureLambdaHandler } from '@aws-lambda-powertools/tracer/middleware';
import mysql from 'mysql';
import { promisify } from 'node:util';
import originalCaptureMySQL from 'aws-xray-sdk-mysql';
import middy from '@middy/core';

const logger = new Logger({ logLevel: 'debug' });
const tracer = new Tracer();
// @ts-ignore - see https://github.com/aws/aws-xray-sdk-node/issues/450 as for why it's needed
const captureMySQL: typeof originalCaptureMySQL.captureMySQL =
  originalCaptureMySQL;
captureMySQL(mysql);

const connection = mysql.createConnection({
  host: process.env.DB_HOST,
  user: process.env.DB_USER,
  password: process.env.DB_PASS,
  database: process.env.DB_DB,
  port: process.env.DB_PORT,
});

export const handler = middy()
  .use(captureLambdaHandler(tracer))
  .handler(async () => {
    try {
      const query = promisify(connection.query).bind(connection);
      const results = await query('SELECT 1 + 1 AS solution'); // this call is now traced
      logger.debug('Results', { results });
    } catch (error) {
      logger.error('Error', { error });
      throw error;
    }
  });
```

While I have tested this with MySQL (see the trace [here](https://github.com/aws-powertools/powertools-lambda-typescript/issues/749#issuecomment-2631569075)), I haven't done the same with PostgreSQL but I expect it to work since the implementation of the patching is very similar on the AWS X-Ray SDK side.
